firmware/library/L1_Drivers/pin_configure.hpp

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <assert.h>
+#include <cstdio>
+
+#include "L0_LowLevel/LPC40xx.h"
+#include "L2_Utilities/macros.hpp"
+
+class PinConfigureInterface
+{
+   public:
+    enum PinMode : uint8_t
+    {
+        kInactive = 0,
+        kPullDown,
+        kPullUp,
+        kRepeater
+    };
+    virtual void SetPinFunction(uint8_t function)         = 0;
+    virtual void SetPinMode(PinMode mode)                 = 0;
+    virtual void EnableHysteresis(bool enable_hysteresis) = 0;
+    virtual void SetAsActiveLow(bool set_as_active_low)   = 0;
+    virtual void EnableFastMode(bool enable_fast_mode)    = 0;
+    virtual void SetAsOpenDrain(bool set_as_open_drain)   = 0;
+};
+
+class PinConfigure : public PinConfigureInterface
+{
+   public:
+    enum PinBitMap : uint8_t
+    {
+        kFunction    = 0,
+        kMode        = 3,
+        kHysteresis  = 5,
+        kInputInvert = 6,
+        kSlew        = 9,
+        kOpenDrain   = 10
+    };
+
+    SJ2_PACKED(struct) PinMap_t
+    {
+        using pin_config_register_t = uint32_t;
+        pin_config_register_t _register[6][32];
+    };
+
+    static PinMap_t * pin_map;
+
+    constexpr PinConfigure(const uint8_t _port, const uint8_t _pin)
+        : port(_port), pin(_pin)
+    {
+        assert(_port <= 5 && "Port must be between 0 and 5");
```

Comment:
What happens to assert runtime? What does it do on the firmware?

Reply:
It spits out the following:
> assertion "_port <= 5 && "Port must be between 0 and 5"" failed: file "/home/khalil/Git/SJSU-Dev2/firmware/library/L1_Drivers/pin_configure.hpp", line 56, function: constexpr PinConfigure::PinConfigure(uint8_t, uint8_t)

May want to use our own version of assert, since this one uses up so much space in text.

Reply:
Decided to move from using assert to a constexpr template factory for this.